Add EDE (RFC8914) caching

TCY16 opened the pull request.

Commit:
Add caching EDEs

Commit:
add cached EDE strings

Commit:
use correct edns_list attach function

Commit:
add happy-flow test

wcawijngaards reviewed (commented):
The code has todo items noted in the comments. The allocation of the string needs to be done in the region, when a reply is copied into a region. A reference to the malloced string is a problem when the original gets deleted, and also another new malloced string is a problem that would be leaked when the region is cleared.

Commit:
swap malloc for regional_alloc and add free

Commit:
differentiate between malloc and regional_alloc

Commit:
add @wcawijngaards' review comments

gthess reviewed (commented):
Some initial comments:
- Needs subnetcache (ECS) and cachedb support;
- The long tests related to `unbound-control dump_cache` and `unbound-control load_cache` are currently failling.

Commit:
remove superfluous variables

Commit:
Merge branch 'master' of github.com:NLnetLabs/unbound into features/ede-caching

gthess committed:
Merge branch 'master' into features/ede-caching

gthess committed:
Review for #759:
- Fix SEGFAULT in load_cache control command.
- Change reason_bogus_str to an explicit NULL-terminated string.
- Fix potential memory leak when discarding a message for referrals and
  0 TTL answers.
- Fix reason_bogus initialization in localzone answers.
- reply_info creation in validator is always regional.

gthess committed:
Review for #759:
- Keep EDE information for keys close to key creation.
- Fix inconsistencies between reply and cached EDEs.
- Incorporate EDE caching checks in EDE tests.
- Fix some EDE cases where missing DNSKEY was wrongly reported.

gthess committed:
- Remove redundant checks when attaching EDE to a SERVFAIL answer.

gthess committed:
- For #759: add support for cached EDEs to cachedump

wcawijngaards commented on daemon/cachedump.c:
```diff
@@ -647,11 +651,15 @@ load_msg(RES* ssl, sldns_buffer* buf, struct worker* worker)
 	}
+	/* there may be EDE text after the numbers */
+	if(consumed > 0 && (size_t)consumed < strlen(s))
+		ede_str = s + consumed + 1 /* space */;
```
For both "" strings and NULL strings, there is an ede number a space and then end of line. The `consumed` is then one less then strlen, and `s+consumed+1` points to the end of string. This then later allocates a "" string for every item. But a NULL ptr is really wanted there.

Perhaps use `(size_t)consumed+1 < strlen(s)` as the guard on the if statement, that turns that situation into `ede_str` NULL and then NULL later on. Also "" strings are then turned into NULL, but there are few of those and it does not really change the EDE output, I believe.

gthess replied:
I put space parsing in the format string since it is always there. Does the new code look okay?

wcawijngaards replied:
Yes, the new code looks fine.

wcawijngaards commented on daemon/worker.c:
```diff
@@ -507,11 +507,11 @@ answer_norec_from_cache(struct worker* worker, struct query_info* qinfo,
-				EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
-					worker->scratchpad, LDNS_EDE_DNSSEC_BOGUS, "");
+			/* Attach the cached EDE (RFC8914) */
+			if(worker->env.cfg->ede) {
```
It now always attaches an EDE. It would be good to check that the ede code is not NULL or the string is not NULL, so that it does not attach an empty EDE. Perhaps modify edns_opt_list_append_ede to that it does not append an EDE with no content, or check in the if clause before calling it.

gthess replied:
Added a check for the EDE code for this one; same as the other uses of edns_opt_list_append_ede.

wcawijngaards commented on validator/val_sigcrypt.c:
```diff
@@ -718,9 +718,9 @@ dnskey_verify_rrset(struct module_env* env, struct val_env* ve,
 	verbose(VERB_ALGO, "rrset failed to verify: all signatures are bogus");
 	if(!numchecked) {
-		*reason = "signature missing";
+		*reason = "signatures bogus";
```
I think this should still be signatures missing, and the EDE for MISSING RRSIGS below. But at line 731, after the next else clause, perhaps another else clause could be added. In that else clause, set the ede reason code to EDE_DNSSEC_BOGUS, and also the reason string can be set.

The numchecked refers to the number of signatures inspected, and if that is zero, there are no signatures that matched the keytag. Or no signatures are present.

gthess replied:
I see your point but it is not 100% clear from the [rfc](https://www.rfc-editor.org/rfc/rfc8914.html#name-extended-dns-error-code-10-) if code 10 means there are no RRSIGS at all (we cover that case with `num==0` above) or no rrsigs that we are expecting. I interpreted the former, but I see that the later could also be the case. It mostly boils down to the end user looking at the EDE code, I believe it would be confusing to say rrsigs are missing when at least some rrsigs are there.
If you still think that missing rrsigs is the correct code there I have no problem changing this.

wcawijngaards replied:
The ede code can be the bogus code. I was thinking the error text itself is not improving, when rrsigs that match certain conditions are absent. But, if the text is more germane for many cases, that would be fine for me.

gthess replied:
Ah, if it is the text then I agree!

wcawijngaards commented on validator/validator.c:
```diff
@@ -2151,9 +2146,19 @@ processFinished(struct module_qstate* qstate, struct val_qstate* vq,
+				char* err_str = errinf_to_str_bogus(qstate);
+				if(err_str) {
+					size_t err_str_len = strlen(err_str);
+					log_info("%s", err_str);
```
The debug print out probably needs to be removed.
```suggestion
```

gthess replied:
I see that this log_info is also on master. Are you sure?

wcawijngaards replied:
You are right, there is a check on the log level. It should not be removed.

wcawijngaards reviewed (approved):
Code looks good. Caching errors could be very helpful for debugging attempts.

gthess committed:
Address review comments for #759:
- Decrease allocations for "" EDE strings when loading the cachedump.
- Check for existence of EDE code before attaching.

gthess committed:
Address review comments for #759:
- Clear error text when an expected signature is missing.

gthess commented:
Thanks @wcawijngaards for the review!
I believe the easiest way (because of all the changes here) to now include the feature for cachedb and ecs is to bring @TCY16's relevant commits here from #790.

gthess committed:
- For #759: fix doc string.
